webapp/src/main/java/com/box/l10n/mojito/smartling/SmartlingClient.java

```diff
@@ -0,0 +1,188 @@
+        return new HttpEntity<>(payload, getHttpHeadersForJsonWithAuthorization());
+    }
+
+    public void authenticate() throws SmartlingClientException {
+        Map<String, Object> payload = getPayloadMapAuthentication();
+        AuthenticationResponse authenticationResponse;
+        try {
+            authenticationResponse = postForAuthentication(payload, AUTH_API_V2_AUTHENTICATE);
+        } catch (SmartlingClientException e) {
+            throw new SmartlingClientException(e.getMessage());
+        }
+        if (authenticationResponse.isAuthenticationErrorResponse()) {
+            throw new SmartlingClientException("Error authenticating with Smartling API, please check your credentials");
+        } else if (!authenticationResponse.isSuccessResponse()) {
+            throw new SmartlingClientException(authenticationResponse.getErrorMessage());
+        } else {
+            Instant now = Instant.now();
+            lastAuthentication = now;
+            lastRefresh = now;
+            nextAuthentication = now.plus(Duration.ofHours(authenticationDurationHours));
+            nextRefresh = now.plus(Duration.ofMinutes(refreshDurationMinutes));
+            authToken = authenticationResponse.getResponse().getData().getAccessToken();
+            refreshToken = authenticationResponse.getResponse().getData().getRefreshToken();
+        }
+    }
+
+    AuthenticationResponse postForAuthentication(Map<String, Object> payload, String authApiV2) throws SmartlingClientException {
+        HttpEntity<Map<String, Object>> httpEntity = getHttpEntityForPayload(payload);
+        String postForObject = restTemplate.postForObject(
+                authApiV2, httpEntity, String.class);
+        AuthenticationResponse authenticationResponse;
+        try {
+            authenticationResponse = objectMapper.readValue(postForObject, AuthenticationResponse.class);
+        } catch (IOException e) {
+            logger.error(e.getMessage());
+            throw new SmartlingClientException(e.getMessage());
+        }
+        return authenticationResponse;
+    }
+
+    public void refresh() throws SmartlingClientException {
+        Instant now = Instant.now();
+        Instant nextCalculatedRefresh = now.plus(Duration.ofMinutes(refreshDurationMinutes));
+        if (nextRefresh.isAfter(now) && nextAuthentication.isAfter(now)) {
+            return;
+        } else if (nextCalculatedRefresh.isAfter(nextAuthentication)) {
+            authenticate();
+            return;
+        }
+        Map<String, Object> payload = getPayloadMapRefresh();
+        AuthenticationResponse authenticationResponse;
+        try {
+            authenticationResponse = postForAuthentication(payload, AUTH_API_V2_REFRESH);
+        } catch (SmartlingClientException e) {
+            throw new SmartlingClientException(e.getMessage());
+        }
+
+        if (authenticationResponse.isAuthenticationErrorResponse()) {
+            authenticate();
```

Comment:
what's the different between those 3 error cases? why do we sometime re-authenticate and sometimes throw an exception?

Reply:
1. authentication error means we should authenticate 
2. any other response code that is not success and that is not related to authentication should be surfaced 
3. not authentication error and not not success code means it is a success code and we should set the refresh token.